File: vbond/vbond/report/revenue_vs_cogs_report/revenue_vs_cogs_report.py
# Copyright (c) 2025, GreyCube Technologies and contributors
# For license information, please see license.txt
import frappe
from erpnext.accounts.report.profit_and_loss_statement.profit_and_loss_statement import (
    get_period_list,
    get_report_summary,
    get_chart_data,
    get_data,
    get_columns,
    get_net_profit_loss,
    execute as _execute
)


def execute(filters=None):
    # return _execute(filters)
    period_list = get_period_list(
        filters.from_fiscal_year,
        filters.to_fiscal_year,
        filters.period_start_date,
        filters.period_end_date,
        filters.filter_based_on,
        filters.periodicity,
        company=filters.company,
    )

    income = get_data(
        filters.company,
        "Income",
        "Credit",
        period_list,
        filters=filters,
        accumulated_values=filters.accumulated_values,
        ignore_closing_entries=True,
        ignore_accumulated_values_for_fy=True,
    )

    expense = get_data(
        filters.company,
        "Expense",
        "Debit",
        period_list,
        filters=filters,
        accumulated_values=filters.accumulated_values,
        ignore_closing_entries=True,
        ignore_accumulated_values_for_fy=True,
    )

    if filters.get("income_account"):
        tmp = list(
            filter(lambda x: x.get("account") == filters.get("income_account"), income)
        )
        tmp[0]["parent_account"] = income[0]["account"]
        for d in income:
            for period in period_list:
                d[period.key] = tmp[0][period.key]
        income = [income[0]] + tmp + income[-2:]

    if filters.get("expense_account"):
        tmp = list(
            filter(
                lambda x: x.get("account") == filters.get("expense_account"), expense
            )
        )
        tmp[0]["parent_account"] = expense[0]["account"]
        for d in expense:
            for period in period_list:
                d[period.key] = tmp[0][period.key]
        expense = [expense[0]] + tmp + expense[-2:]

    net_profit_loss = get_net_profit_loss(
        income, expense, period_list, filters.company, filters.presentation_currency
    )

    data = []
    data.extend(income or [])
    data.extend(expense or [])
    if net_profit_loss:
        data.append(net_profit_loss)

    columns = get_columns(
        filters.periodicity, period_list, filters.accumulated_values, filters.company
    )

    
    currency = filters.presentation_currency or frappe.get_cached_value(
		"Company", filters.company, "default_currency"
	)
    chart = get_chart_data(filters, columns, income, expense, net_profit_loss, currency)

    report_summary = get_report_summary(
        period_list,
        filters.periodicity,
        income,
        expense,
        net_profit_loss,
        currency,
        filters,
    )

    return columns, data, None, chart, report_summary

File: vbond/vbond/report/revenue_vs_cogs_report/revenue_vs_cogs_report.js
// Copyright (c) 2025, GreyCube Technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Revenue vs COGS Report"] = $.extend({}, erpnext.financial_statements);

erpnext.utils.add_dimensions("Revenue vs COGS Report", 10);

frappe.query_reports["Revenue vs COGS Report"]["filters"].push({
	fieldname: "selected_view",
	label: __("Select View"),
	fieldtype: "Select",
	options: [
		{ value: "Report", label: __("Report View") },
		{ value: "Growth", label: __("Growth View") },
		{ value: "Margin", label: __("Margin View") },
	],
	default: "Report",
	reqd: 1,
});

frappe.query_reports["Revenue vs COGS Report"]["filters"].push({
	fieldname: "accumulated_values",
	label: __("Accumulated Values"),
	fieldtype: "Check",
	default: 1,
});

frappe.query_reports["Revenue vs COGS Report"]["filters"].push({
	fieldname: "include_default_book_entries",
	label: __("Include Default FB Entries"),
	fieldtype: "Check",
	default: 1,
});

frappe.query_reports["Revenue vs COGS Report"]["filters"].push({
	"fieldname": "income_account",
	"label": __("Income Account"),
	"fieldtype": "Link",
	"options": "Account"
});

frappe.query_reports["Revenue vs COGS Report"]["filters"].push({
	"fieldname": "expense_account",
	"label": __("Expense Account"),
	"fieldtype": "Link",
	"options": "Account"
});
